internal/kms/kms.go

```
@@ -269,6 +273,254 @@ func (k *Kms) RotateKeys(ctx context.Context, scopeId string, opt ...Option) err
+	slices.SortFunc(foundKey.Versions, func(i, j wrappingKms.KeyVersion) bool {
+		return i.Version < j.Version
+	})
+	if foundKey.Versions[len(foundKey.Versions)-1].Id == keyVersionId {
+		// Attempted to destroy currently active key
+		return false, errors.New(ctx, errors.InvalidParameter, op, "cannot destroy a currently active key version")
+	}
+	switch foundKey.Purpose {
+	case wrappingKms.KeyPurpose(KeyPurposeOplog.String()):
+		return false, errors.New(ctx, errors.InvalidParameter, op, "oplog key versions cannot be destroyed")
+	case wrappingKms.KeyPurposeRootKey:
+		// Simple case, just rewrap and destroy synchronously
+		if err := k.underlying.RewrapKeys(ctx, scopeId); err != nil {
+			return false, errors.Wrap(ctx, err, op)
+		}
+		if err := k.underlying.RevokeKeyVersion(ctx, keyVersionId); err != nil {
+			return false, errors.Wrap(ctx, err, op)
+		}
+		return true, nil
+	}
+
+	tablesNeedingRewrapping := 0
+	if _, err := k.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{}, func(r db.Reader, w db.Writer) error {
+		var tables []*DataKeyVersionDestructionJobRunAllowedTableName
+		if err := r.SearchWhere(ctx, &tables, "1=1", nil, db.WithLimit(-1)); err != nil {
+			return errors.Wrap(ctx, err, op, errors.WithMsg("failed to look up allowed table names"))
+		}
+		// Start by adding any job runs necessary by looking up the number of
+		// rows encrypted with the key that's being destroyed.
+		for _, table := range tables {
+			rows, err := r.Query(ctx, fmt.Sprintf(findAffectedRowsForKeyQueryTemplate, table.GetTableName()), []interface{}{keyVersionId})
+			if err != nil {
+				errors.Wrap(ctx, err, op, errors.WithMsg("failed to get affected rows for %q", table.GetTableName()))
+			}
+			defer rows.Close()
+			var numRows uint
+			for rows.Next() {
+				if err := r.ScanRows(ctx, rows, &numRows); err != nil {
+					errors.Wrap(ctx, err, op, errors.WithMsg("failed to scan number of rows for %q", table.GetTableName()))
```

these two errors are not returned?

---

Hm, oops!